[TOC levels=1-3]: #

# Table of Contents
- [Table of Contents](#table-of-contents)
- [Name](#name)
- [Purpose](#purpose)
- [Data downloading](#data-downloading)
	- [Sequencing data](#sequencing-data)
	- [Reference genome data](#reference-genome-data)
- [Quality control and trimming](#quality-control-and-trimming)
- [Methylation analysis](#methylation-analysis)
	- [Genome indexing](#genome-indexing)
	- [Read alignment](#read-alignment)
	- [Aligned reads deduplication](#aligned-reads-deduplication)
	- [Methylation information extracting](#methylation-information-extracting)
- [Downstream analysis](#downstream-analysis)
	- [Input data preparation](#input-data-preparation)
	- [DML/DMR detection](#dmldmr-detection)
- [Practical methylation information analysis](#practical-methylation-information-analysis)
- [Reference](#reference)
- [Author](#author)

# Name
5hmC finding procedure - bioinfo procedures of finding the 5hmC DMLs in mouse samples based on the ACE-seq data coming from [pushlished paper](https://www.nature.com/articles/nbt.4204).     
该文章提出了一个区分5mC和5hmC的方法——ACE-Seq


# 背景补充：  
* ACE-Seq：羟甲基化检测    

DNA序列中胞嘧啶（C）5´碳位结合一个甲基基团而转变成5mC的现象称为DNA甲基化。在5mC去甲基化的过程中，其会被氧化成5hmC，这种新的修饰称为DNA羟甲基化5-hydroxymethylcytosine。  

借助重亚硫酸盐（Bisulfite）转换的方法被认为是金标准，该化学试剂会将未甲基化的C进行转化，胞嘧啶（C）转换为尿嘧啶（U），后续的PCR扩增过程U会被胸腺嘧啶（T）替代，而，甲基化的C则不受影响。   


因为5hmC是5mC去甲基化的中间产物，所以对于多细胞来说，同一个C位点上不同细胞往往会存在5mC或5hmC不同修饰，而5mC与5hmC对基因表达的调控作用又完全不同：启动子区的DNA甲基化对基因表达的有抑制作用，DNA羟甲基化则会促进基因的表达。因此，对5mC及5hmC的水平进行精确地定量并分别研究十分有必要。   


APOBEC-coupled epigenetic sequencing (ACE-seq)是一种全新的检测DNA羟甲基化的技术，由美国宾夕法尼亚大学Emily K Schutsky等在2018年发表在Nature Biotechnology杂志上。不同于传统的BS-Seq，该技术使用AID/APOBEC家族DNA脱氨酶APOBEC3A (A3A)代替化学脱氨，保证了DNA的完整性。A3A能够特异地脱去C及5mC的氨基使其变为U，而5hmC则因为不被该酶识别，测序时仍被检测为C。这样，便将5hmC与C及5mC区分开来。

首先，利用T4 β-葡糖基转移酶（βGT）将5hmC转化成5ghmC，5ghmC不会被A3A酶脱氨基；然后，利用A3A酶特异去除C和5mC的氨基，使其转变成U，而5hmC保持为C，从而将5hmC和C/5mC区分开来。  






# Purpose
* Learn the methylation data analysis procedure based on software tool [`bismark`](https://github.com/FelixKrueger/Bismark) and [`DSS`](http://bioconductor.org/packages/release/bioc/html/DSS.html) package.
* Get familiar with the Linux sever workspace and the basic bioinfo data analysis protocol.

# Data downloading
* The sequencing data comes from the NBT paper and the GEO accession number is [GSE116016](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116016)    



* The Mus musculus reference genome data is downloaded from [Ensembl](https://www.ensembl.org/info/data/ftp/index.html)   


* All data are downloaded via FTP address by a lightweight downloading tool [aria2](https://aria2.github.io/)
* Detailed information about data downloading ==> [Blog](https://www.jianshu.com/u/3fcc93cd84c1)
* Data used in this sample procedure:
    - Wild type: SRR7368841 & SRR7368842
    - TetTKO type: SRR7368845
    
## Sequencing data
Open browser, vist [EBI-ENA search page](https://www.ebi.ac.uk/ena) and search for the GEO accession. On the [result page](https://www.ebi.ac.uk/ena/data/view/PRJNA476795), we could get the expected FTP address. Use `aria2` to download the data and we could get `.fastq.gz` files.

```bash
mkdir -p $HOME/NBT_repeat/data/seq_data/WT_mESC_rep1
mkdir -p $HOME/NBT_repeat/data/seq_data/TetTKO_mESC_rep1
cd $HOME/NBT_repeat/data/seq_data/

aria2c -d ./WT_mESC_rep1/ -Z ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR736/001/SRR7368841/SRR7368841.fastq.gz ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR736/002/SRR7368842/SRR7368842.fastq.gz
aria2c -d ./TetTKO_mESC_rep1/ -Z ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR736/005/SRR7368845/SRR7368845.fastq.gz
```

```
建立文件夹 d/methylation/data/sequence
cd /mnt/d/methylation/data/sequence
cat >1.txt <<EOF
SRR7368841
SRR7368842
SRR7368845
EOF


prefetch --option-file 1.txt

下载到了 ~/data/sra文件夹下，因为某些测序文件太大，需要重新单独下载
@ ~/data/sra$  prefetch  --max-size 30GB SRR7368844  SRR7368847  SRR7368851


~/data/sra$ ls

parallel -j 4 "    
    fastq-dump --split-3 --gzip {1}  -O /mnt/d/methylation/data/sequence   
" ::: $(ls *.sra)     

# ls *.sra代表，列举出任何以.sra结尾的文件
--gzip 将转换出的fastq文件以gz格式输出，可以节省空间
--split-3 把pair-end测序分成两个文件输出，可用于双端测序转化为两个文件，本文举例为单端测序，删掉不影响
-O 输出文件夹名，不加直接放在该文件夹


# 删除sra文件
$ rm *.sra

gzip -d -c SRR7368841.fastq.gz | head -n 20

```
## Reference genome data 
Open browser, visit [ensembl download page](https://www.ensembl.org/info/data/ftp/index.html), and find the DNA `.fasta` file. On the [result page](ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/), we could get the expected FTP address.

```bash
mkdir -p $HOME/NBT_repeat/data/genome_data/
cd $HOME/NBT_repeat/data/genome_data/

# Basic command line download
aria2c -d ./ -Z ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.1.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.2.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.3.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.4.fa.gz
aria2c -d ./ -Z ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.5.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.6.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.7.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.8.fa.gz
aria2c -d ./ -Z ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.9.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.10.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.11.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.12.fa.gz
aria2c -d ./ -Z ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.13.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.14.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.15.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.16.fa.gz
aria2c -d ./ -Z ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.17.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.18.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.19.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.X.fa.gz ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.Y.fa.gz 

# Alternative option: script download
sh $HOME/Scripts/shell/genome_data_download.sh $HOME/NBT_repeat/data/genome_data/
```

和直接从ensembl进去选择大鼠是一样的  

测序平台 Platforms (1)：GPL19057	Illumina NextSeq 500 (Mus musculus)    
只需要下载大鼠肌肉组织的基因组数据，且挑选一个较短的染色体(19号)进行下面步骤     




```bash
cd /mnt/d/methylation/data/gene

aria2c -d ./ http://ftp.ensembl.org/pub/release-107/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.chromosome.19.fa.gz  

gzip -d Mus_musculus.GRCm39.dna.chromosome.19.fa.gz

mv Mus_musculus.GRCm39.dna.chromosome.19.fa   Mus.chr19.fa


# 每一条primary_assembly的名称后面还跟了一些描述信息，这些描述信息就是当前组装版本，长度等等信息，
# 但是这个信息会妨碍后面写脚本统计或者一些分析，所以这里最好去掉 
# mv Mus.chr19.fa Mus.chr19.raw.fa
# 然后去除染色体编号后的描述信息
# cat Mus.chr19.raw.fa | perl -n -e 'if(m/^>(.+?)(?:\s|$)/){ print ">$1\n";}else{print}' > Mus.chr19.fa
# 删除
# rm Mus.chr19.raw.fa
```


# Quality control and trimming
Performing some quality control is highly recommended for all high throughput sequencing to tell straight whether the dataset is of good quality or whether there
were any fundamental problems with either the library or the sequencing itself.

Here, we use [FastQc](www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [Trim Galore](www.bioinformatics.babraham.ac.uk/projects/trim_galore/) to do the quality control and adapter trimming respectively. Trimming reports could be found in sub-directory [`trimming-reports/`](trimming-reports/).

```bash
mkdir /mnt/d/methylation/output/fastqc
cd /mnt/d/methylation/data/sequence

# quality control to see the quality of the raw seq-data
fastqc -t 3 /mnt/d/methylation/data/sequence/*.fastq.gz -o /mnt/d/methylation/output/fastqc

# 整合质控结果
cd /mnt/d/methylation/output/fastqc
multiqc .
``` 
fastqc发现，其碱基比例严重失调（C几乎没有），因为本样本测序测定了处理后的甲基化水平，C已经被变成U/T，只有5hmC仍被识别为C，所以不算错误；  
adapter含量高，需要去除接头  


```bash

# 剔除接头以及测序质量差的碱基
# quality and adapter trimming, followd by fastqc operation质控和去除接头两个步骤一起
mkdir -p ../adapter/

trim_galore -o /mnt/d/methylation/output/adapter/ --fastqc /mnt/d/methylation/data/sequence/*.fastq.gz

# 整合质控结果
cd /mnt/d/methylation/output/adapter
multiqc .
```

`FastQc` used options:
* `-t/--threads <int>`: Specifies the number of files which can be processed simultaneously.

`Trim Galore` used options:
* `-o/--output_dir <DIR>`: If specified all output will be written to this directory instead of the current directory. If the directory doesn't exist it will be created for you.  
* `--fastqc`: Run FastQC in the default mode on the FastQ file once trimming is complete.

# Methylation analysis
After the quality control and adapter trimming, BS-seq data analysis protocal based on [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) could be followed to do the methylation analysis for the ACE-seq data as well.  

在质量控制和adapter修整之后，也可以按照基于 Bismarck 的 BS-seq 数据分析协议对 ACE-seq 数据进行甲基化分析。  


bismark的分析分为三步：  


* 基因组索引构建：需要将基因组进行相应的转换，`模拟bisulfite处理后基因组`，这样才能用于比对。在每个新物种比对之前，都需要进行该处理。bismark内部默认调用bowite软件进行比对。  


* 序列比对：输入为下机序列、基因组和比对参数，bismark会输出比对结果和甲基化检测的结果，默认格式为BAM，同时还会有一个统计报告。  


* 甲基化位点提取：这一步是可选的，会利用bismark的比对结果来获得甲基化信息。这一步会把甲基化分为不同的类别（CG，CHG和CHH），获得链特异性信息，并且提供过滤参数；也可以对甲基化信息进行调整，或者进行更多的深入分析。  


链接：https://www.jianshu.com/p/877a5716c24a  


## Genome indexing 
Before alignments can be carried out, the genome of interest needs to be bisulfite converted *in-silico* and indexed. Based on `Bismark`, we can use both `Bowtie` and `Bowtie2` to do the indexing work.   

在进行比对之前，需要对感兴趣的基因组进行亚硫酸氢盐在计算机上进行转换并编制索引。 基于 Bismark，我们可以同时使用 Bowtie 和 Bowtie2 来做索引工作。
 
* 代码  

```bash 
#下载bismark
https://www.bioinformatics.babraham.ac.uk/projects/download.html#bismark
or 
brew install bismark
cd /mnt/d/methylation/data/gene/
bismark_genome_preparation --bowtie2 /mnt/d/methylation/data/gene/  --verbose # 输出log信息
``` 

* 输出文件:

```
bismark会生成两个目录，一个是C->T转换的基因组，一个是G->A转换的基因组共两套基因组，之后可以用bowtie2-build来建立索引。

特别注意，索引路径中必需包含一个基因组文件。
```


`bismark_genome_preparation` used options:
* `--bowtie2`: This will create bisulfite indexes for Bowtie 2. (Default: ON).

## Read alignment
The core of the methylation data analysis procedure is to align the sequencing reads to the reference genome, and it is assumed that all data have been quality and adapter trimmed. The alignment reports could be found in sub-directory [`bismark-alignment-reports/`](bismark-alignment-reports/).  


甲基化数据分析程序的核心是将测序读数与参考基因组对齐，并假设所有数据都经过质量和adapter修整。 比对报告可以在子目录中找到[`bismark-alignment-reports/`](bismark-alignment-reports/).    



* 比对的过程需要提供两个信息：  

```
输入目录：已经建立好的索引目录，即基因组索引步骤中的索引目录
输入文件：测序reads，fastq格式
```
* 输出包括：
```
比对的结果，BAM文件，可以利用SeqMonk进行可视化；也可以用于继续的去除dup等
报告文件：包含各种统计信息和甲基化统计结果。报告展示了唯一比对上的read对数目等信息（Final Alignment report部分）以及有reads支持发生甲基化的C的数目统计（Final Cytosine Methylation Report部分），注意这里并非真正意义上的甲基化C，仍需要下游的分析。

比对输出路径中包含一个比对后的文件（BAM格式）和一个比对报告文件以及未比对到参考基因组上的reads文件（--un参数）。

若测序数据为双端数据文件中会有PE或pe标签，若为单端数据则会有SE或se字符加以区别。

针对链特异性文库，若进行单端比对read1使用bismark单端参数即可，
read2需要添加--pbat参数；而非链特异性文库--pbat参数的使用则相反。

通常，若某物种比对率较低，可先进行双端比对同时输出未必对的reads，随后使用未比对的read1和read2分别进行单端比对并合并输出的BAM文件，作为下游的输入。
```

* 代码  

```bash
genome_path="/mnt/d/methylation/data/gene/" # 已经index的基因组文件
cd /mnt/d/methylation/output/adapter  # 使用已经去过接头的测序文件来比对
mkdir /mnt/d/methylation/output/bismark_align_results


# read alignment
bismark -o /mnt/d/methylation/output/bismark_align_results \
 --parallel 4 --bam --genome_folder ${genome_path} ./*.fq.gz

# 2>../bismark_results/{1}.log # 标准输出输入到相对应的文件中，已经有结果文件了
```

* 不重要 

```bash

# 在比对过程中，一个样品产生了四个文件，原本以为需要合并bam文件，结果不需要
# 计算alignments数，不一定能成

cd /mnt/d/methylation/output/bismark_align_results

cat file_name.txt
SRR7368841_trimmed.fq.gz.temp.1_bismark_bt2
SRR7368841_trimmed.fq.gz.temp.2_bismark_bt2
SRR7368841_trimmed.fq.gz.temp.3_bismark_bt2
SRR7368841_trimmed.fq.gz.temp.4_bismark_bt2

cat file_name.txt | while read line 
do
export alignment_number=$(samtools view -c ${line}.bam)
echo ${line} alignment_number ${alignment_number}
done

# merge the 4 result .bam file
samtools merge -n SRR7368841_trimmed.fq.gz.bismark_bt2.bam \
-b ${file_name}.bam  

# 再次检查  
```

* `bismark` 重要参数:  
```
bismark 
-N：设置seed中允许的最大错配数，可取0或1，默认为0。值越高，对齐速度越慢，灵敏度越高。
-L：设置seed长度，最大值为32，默认为20。值越高，对齐速度越快，灵敏度越低。
--quiet：不输出比对流程信息
--un：过滤多处匹配的reads
--ambiguous：多处匹配reads信息独立记录
--sam/--bam：输出SAM格式，与--parallel不兼容/输出BAM格式，可调整--parallel
--bowtie2/--hisat2 ：调用bowtie2/hisat2，默认bowtie2
--path_to_bowtie2/--path_to_hisat2 ：bowtie2/hisat2所在文件夹的全路径
-o/--output_dir ：输出文件的全路径
--samtools_path：samtools所在文件夹的全路径
--prefix：指定输出文件的前缀
--q/--fastq：输入文件为FastQ
-f/--fasta：输入文件为FastA
--phred33-quals/--phred64-quals：指定FastQ文件的质量分数格式，默认为phred33
-1/-2：双端测序文件
genome_folder：包含未修改的参考基因组和bismark_genome_preparation创建的子文件夹(CT_conversion/和GA_conversion/）的文件夹的路径

链接：https://www.jianshu.com/p/0aa453ba9476
```


* `-o/--output_dir <dir>`: Write all output files into this directory. By default the output files will be written into the same folder as the input file(s). If the specified folder does not exist, Bismark will attempt to create it first.
* `--parallel <int>`: (May also be `--multicore <int>`) Sets the number of parallel instances of Bismark to be run concurrently. This forks the Bismark alignment step very early on so that each individual Spawn of Bismark processes only every n-th sequence (n being set by --parallel).
* `--genome_folder`: **Arguments**, The path to the folder containing the unmodified reference genome as well as the subfolders created by the `bismark_genome_preparation` script. genome_folder：包含未修改的参考基因组和bismark_genome_preparation创建的子文件夹(CT_conversion/和GA_conversion/）的文件夹的路径.


`samtools` used command and options:
* `cat`: **Command** concatenate BAMs
    - `-o`: output file format



## Aligned reads deduplication
Mammalian genomes are so huge that it is rather unlikely to encounter several genuinely independent fragments which align to the very same genomic position. It is much more likely that such reads are a result of PCR amplification. For large genomes, removing duplicate reads is therefore a valid route to take.  

哺乳动物基因组如此庞大，以至于不太可能遇到几个对齐到同一位置的独立片段，它们与相同的基因组位置对齐。 这些读数更有可能是 PCR 扩增的结果。 因此，对于大型基因组，去除重复读数是一条有效的途径。由于PCR的扩增，可能会存在比对到基因组相同位置的序列，重复测序的序列需要被去除。如果是RRBS文库，建议不要进行这一步骤。  



The deduplication step could be finished through the `deduplicate_bismark` script, and the deduplication report could be found in sub-directory [`deduplication-reports/`](##deduplication-reports/).去重步骤可以通过`deduplicate_bismark`脚本完成，去重报告可以在子目录[`deduplication-reports/`](##deduplication-reports/)中找到。

* 输入与输出  

```
输入文件：上一步比对好的bam文件（包含甲基化信息等）
输出文件：
test.file.R1_bismark_bt2_pe.deduplicated.sam  去重了的sam文件
test.file.R1_bismark_bt2_pe.deduplication_report.txt
```

* 代码  

```bash
mkdir -p /mnt/d/methylation/output/deduplicated_results

# aligned reads deduplication
deduplicate_bismark --bam \
--output_dir /mnt/d/methylation/output/deduplicated_results/ \
/mnt/d/methylation/output/bismark_align_results/*.bam
```

* 重要参数    

```bash
deduplicate_bismark 

--sam/--bam：删除 Bismark 对齐产生的SAM/BAM 文件中的重复数据，建议用于WGBS，但不建议应用于RRS (reduced representation shotgun)，如 RRBS、amplicon or target enrichment libraries。
-p/--paired ：前一步双端数据产生的结果文件
-s/--single：前一步单端数据产生的结果文件
--samtools_path：samtools所在文件夹的全路径
--output_dir：输出文件夹路径
--multiple：指定输入文件都作为一个样本处理，连接在一起进行重复数据删除。

对SAM文件使用Unix“cat”，对BAM文件使用“samtools cat”。所有输入文件的格式必须相同。默认情况下，标头取自要连接的第一个文件。

链接：https://www.jianshu.com/p/0aa453ba9476
```
`deduplicate_bismark` used options:
* `--bam`: The output will be written out in BAM format instead of the default SAM format.
* `--output_dir`: Output directory, either relative or absolute. Output is written to the current directory if not specified explicitly.

## Methylation information extracting


The `bimark_methylation_extractor` script in the `Bismark` package could extract the methylation information from the alignment result files and act as the endpoint of the `Bismark` package. In addition, the methylation information could be easily transfered into other format facilitating the downstream analysis in this script.   


`Bismark`包中的`bimark_methylation_extractor`脚本可以从比对结果文件中提取甲基化信息，作为`Bismark`包的端点。 此外，甲基化信息可以很容易地转换为其他格式，便于在此脚本中进行下游分析。  

可以使用--bedGraph，来产生bedgraph和coverage的文件；--count可以产生count文件，输出每个CpG位点的深度文件。  


```bash
mkdir /mnt/d/methylation/output/extracting_results
genome_path= "/mnt/d/methylation/data/gene/"

# methylation information extracting
bismark_methylation_extractor --single-end --gzip --parallel 4 --bedGraph \
--cytosine_report --genome_folder ${genome_path} \
-o /mnt/d/methylation/output/extracting_results/  /mnt/d/methylation/output/deduplicated_results/*.bam
```

`bismark_methylation_extractor` used options:
```
bismark_methylation_extractor
    --pair-end         指定双端数据
    --comprehensive    输出CHG CHH CpG的甲基化信息
    --no-overlap       去除reads重叠区域的bias
    --bedGraph         输出bedGraph文件
    --counts           每个C上甲基化reads和非甲基化reads的数目
    --report           一个甲基化summay
    --cytosine_report  输出全基因组所有CpG
    --genome_folder    <path_to_reference_genome>
    input.bam          输入文件
    -o >output_dir>    输出路径
```
* `--gzip`: The methylation extractor files (CpG_OT_..., CpG_OB_... etc) will be written out in a GZIP compressed form to save disk space. This option is also passed on to the genome-wide cytosine report.
* `--parallel <int>`: May also be `--multicore <int>`. Sets the number of cores to be used for the methylation extraction process.
* `--bedGraph`: After finishing the methylation extraction, the methylation output is written into a sorted bedGraph file that reports the position of a given cytosine and its methylation state.完成甲基化提取后，将甲基化输出写入排序的 bedGraph 文件，该文件报告给定胞嘧啶的位置及其甲基化状态。  

* `--cytosine_report`: After the conversion to bedGraph has completed, the option `--cytosine_report` produces a genome-wide methylation report for all cytosines in the genome. By default, the output uses 1-based chromosome coordinates (zero-based start coords are optional) and reports CpG context only (all cytosine context is optional). The output considers all Cs on both forward and reverse strands and reports their position, strand, trinucleotide content and methylation state (counts are 0 if not covered). 转换为 bedGraph 完成后，选项 `--cytosine_report` 为基因组中的所有胞嘧啶生成全基因组甲基化报告。 默认情况下，输出使用从 1 开始的染色体坐标（从零开始的起始坐标是可选的）并仅报告 CpG 上下文（所有胞嘧啶上下文都是可选的）。 输出考虑正向和反向链上的所有 C，并报告它们的位置、链、三核苷酸含量和甲基化状态（如果未覆盖，则计数为 0）。  

* `--genome_folder <path>`: Enter the genome folder you wish to use to extract sequences from (full path only).  
输入希望用于从中提取序列的基因组文件夹.  


* `-o/ --output`: Allows specification of a different output directory (absolute or relative path). If not specified explicitly, the output will be written to the current directory.


* 结果：  
bismark对不同状态的胞嘧啶进行了约定并记录到了BAM文件和中间文件中：  

```
X 代表CHG中甲基化的C
x  代表CHG中非甲基化的C
H 代表CHH中甲基化的C
h  代表CHH中非甲基化的C
Z  代表CpG中甲基化的C
z  代表CpG中非甲基化的C
U 代表其他情况的甲基化C(CN或者CHN)
u  代表其他情况的非甲基化C (CN或者CHN)
```
* bismark_methylation_extractor  

是一个非常重要的输出文件为全基因组胞嘧啶报告文件（*CX_report.txt），该文件是一个非常核心的记录胞嘧啶深度信息的文件，可进行甲基化水平的计算、区域甲基化信号文件的转化、差异甲基化水平的计算等。

```
chr1    14716   +       1       6       CG      CGT
chr1    14717   -       10      1       CG      CGA
chr1    14741   +       2       1       CG      CGG
chr1    14742   -       0       1       CG      CGC


各列信息解释如下：
第1列 : 染色体编号
第2列 : 染色体起始位置
第3列 : 正负链信息
第4列 : 甲基化碱基数目
第5列 : 非甲基化碱基数目
第6列 : CG/CHG/CHH
第7列 : C背景序列

基于该文件可进行单个胞嘧啶位点的甲基化水平（C/C+T）计算，比如第一行：甲基化是水平为1/(1+6)；针对指定区域的甲基化水平计算则取区域内的胞嘧啶的总C/(总C+总T)即可。
```



# Downstream analysis
Based on the methylation information result got from the methylation analysis procedure, we could make some basic downstream analysis work including finding specific locus and detecting differential methylation loci (DML) or differential methylation regions (DMR). Here we use R package [`DSS`](http://bioconductor.org/packages/release/bioc/html/DSS.html) to make the differential methylation analysis.  

根据甲基化分析程序得到的甲基化信息结果，我们可以进行一些基本的下游分析工作，包括寻找特定位点和检测差异甲基化位点（DML）或差异甲基化区域（DMR）。 这里我们使用 R 包 [`DSS`](http://bioconductor.org/packages/release/bioc/html/DSS.html) 进行差异甲基化分析。  



## Input data preparation
`DSS` requires data from each BS-seq like experiment to be summarized into following information for each CG position: chromosome number, genomic coordinate, total number of reads, and number of reads showing methylation.   

`DSS` 要求将来自每个 BS-seq 类实验的数据总结为每个 CG 位置的以下信息：染色体标号、基因组坐标、读取总数和显示甲基化的读取数。   


The required input data could be transfered from the bismark result `.cov` file since the count file contain following columns: chr, start, end, methylation%, count methylated, count unmethylated.  

所需的输入数据可以从 bismark 结果“.cov”文件中传输，因为计数文件包含以下列：chr、开始、结束、甲基化％、甲基化计数、未甲基化计数。


```bash
mkdir -p /mnt/d/methylation/R_analysis/data

cd /mnt/d/methylation/R_analysis/data

# store the file path information
file_path="/mnt/d/methylation/output/deduplicated_results/*.deduplicated.bismark.cov.gz"

# copy the result file to the R analysis folder
cp file_path /mnt/d/methylation/R_analysis

# unzip
gunzip -d /mnt/d/methylation/R_analysis/     
SRX4241790_trimmed_bismark_bt2.deduplicated.bismark.cov.gz

# transfer the .cov file to .txt file
cp /mnt/d/methylation/R_analysis/SRX4241790_trimmed_bismark_bt2.deduplicated.bismark.cov ./WT_data/SRX4241790_methylation_result.txt
cp ./TetTKO_data/SRR7368845_trimmed_bismark_bt2.deduplicated.bismark.cov ./TetTKO_data/SRR7368845_methylation_result.txt

# basic command line environment procedure 
R
```
```r
library(tidyr)
library(dplyr)

file_names <- c("./WT_data/SRX4241790_methylation_result.txt", "./TetTKO_data/SRR7368845_methylation_result.txt")

func_read_file <- function(file_name){
	dir_vec <- strsplit(file_name, split = "/")[[1]]
	len <- length(dir_vec)
	file_prefix = substring(dir_vec[len], 0, nchar(dir_vec[len]) - 4)
	file_save_path = substring(file_name, 0, nchar(file_name) - nchar(dir_vec[len]))
	print(paste("File", file_name, "is being importing and this may take a while..."), sep = "")
	rawdata_df <- read.table(file_name, header = F, stringsAsFactors = F)
	print("Importing file is finished!")
	colnames(rawdata_df) <- c("chr", "start", "end", "methyl%", "methyled", "unmethyled")
	write.table(rawdata_df, paste(file_save_path, file_prefix, "_transfered.txt", sep = ""), row.names = F )
}

lapply(file_names, func_read_file)

q()
```

```bash
# Rscript procedure
Rscript $HOME/Scrpits/R/bismark_result_transfer.R ./WT_data/SRX4241790_methylation_result.txt ./TetTKO_data/SRR7368845_methylation_result.txt
```

## DML/DMR detection
After the input data preparation, we could use `DSS` package to find DMLs or DMRs. The detailed result files of both DML and DMR detection could be found in the sub-directory [`methylation-differential-analysis-reports/`](methylation-differential-analysis-reports/).

```bash
cd $HOME/NBT_repeat/R_analysis/

# basic command line procedure
R
```
```r
library(tidyr)
library(dplyr)
library(DSS)

first_file <- "./WT_data/SRX4241790_methylation_result.txt"
second_file <- "./TetTKO_data/SRR7368845_methylation_result.txt"
file_prefix <- "mm_all_chr"
file_save_path <- "./"

# import data
first_raw_data <- read.table(first_file, header = T, stringsAsFactors = F)
second_raw_data <- read.table(second_file, header = T, stringsAsFactors = F)

# data manipulation to prepare for the BSseq objection
DSS_first_input_data <- first_raw_data %>%
	mutate(chr = paste("chr", chr, sep = "")) %>%
	mutate(pos = start, N = methyled + unmethyled, X = methyled) %>%
	select(chr, pos, N, X)
DSS_second_input_data <- second_raw_data %>%
	mutate(chr = paste("chr", chr, sep = "")) %>%
	mutate(pos = start, N = methyled + unmethyled, X = methyled) %>%
	select(chr, pos, N, X)

# create BSseq object and make the statical test
bsobj <- makeBSseqData(list(DSS_first_input_data, DSS_second_input_data), c("S1", "S2"))
dmlTest <- DMLtest(bsobj, group1 = c("S1"), group2 = c("S2"), smoothing = T)

# dmls detection
dmls <- callDML(dmlTest, p.threshold = 0.001)
# dmrs detection
dmrs <- callDMR(dmlTest, p.threshold=0.01)

# output the results
write.table(dmlTest, paste(file_save_path, file_prefix, "_DSS_test_result.txt", sep = ""), row.names = F)
write.table(dmls, paste(file_save_path, file_prefix, "_DSS_dmls_result.txt", sep = ""), row.names = F)
write.table(dmrs, paste(file_save_path, file_prefix, "_DSS_dmrs_result.txt", sep = ""), row.names = F)
```

```bash
# Rscript procedure
Rscript $HOME/Scripts/R/DSS_differ_analysis.R ./WT_data/SRX4241790_methylation_result.txt ./TetTKO_data/SRR7368845_methylation_result.txt mm_chr_all ./
```

# Practical methylation information analysis
Here we present a simple method to take advantage of the methylation information we got from the orthodox methylation analysis procedure.


# Reference
* [Aria2 Manual](https://aria2.github.io/manual/en/html/index.html)
* [Bismark User Guide](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html)
* [DSS package Manual](http://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.pdf)
* [Bismark软件使用入门](https://www.jianshu.com/p/877a5716c24a)


# 遇到的问题

* 最大的问题  

[https://blog.csdn.net/weixin_39471848/article/details/121566810](https://blog.csdn.net/weixin_39471848/article/details/121566810)  


```bash
xuruizhi@DESKTOP-HI65AUV:~$ bismark

perl: error while loading shared libraries: libcrypt.so.1: cannot open shared object file: No such file or directory

xuruizhi@DESKTOP-HI65AUV:~$ which perl
/home/linuxbrew/.linuxbrew/bin/perl

xuruizhi@DESKTOP-HI65AUV:~$ ldd /home/linuxbrew/.linuxbrew/bin/perl
        linux-vdso.so.1 (0x00007ffde0ffc000)
        libperl.so => /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/x86_64-linux-thread-multi/CORE/libperl.so (0x00007f5470200000)
        libc.so.6 => /home/linuxbrew/.linuxbrew/lib/libc.so.6 (0x00007f546ffef000)
        libpthread.so.0 => /home/linuxbrew/.linuxbrew/lib/libpthread.so.0 (0x00007f5470958000)
        libdl.so.2 => /home/linuxbrew/.linuxbrew/lib/libdl.so.2 (0x00007f5470953000)
        libm.so.6 => /home/linuxbrew/.linuxbrew/lib/libm.so.6 (0x00007f547086f000)
        /home/linuxbrew/.linuxbrew/lib/ld.so => /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/lib64/ld-linux-x86-64.so.2 (0x00007f547095f000)
        # 没有找到依赖包
        libcrypt.so.1 => not found

# 查找丢失的共享库信息       
xuruizhi@DESKTOP-HI65AUV:~$ find ~ -name  libcrypt.so.1
/home/xuruizhi/libcrypt.so.1

xuruizhi@DESKTOP-HI65AUV:~$ which perl
/home/linuxbrew/.linuxbrew/bin/perl


# 建立软链接
xuruizhi@DESKTOP-HI65AUV:~$ ln -s /home/xuruizhi/libcrypt.so.1 /home/linuxbrew/.linuxbrew/bin/../lib/libcrypt.so.1

# 检查软链接是否建立成功
xuruizhi@DESKTOP-HI65AUV:~$ ldd /home/linuxbrew/.linuxbrew/bin/perl
        linux-vdso.so.1 (0x00007ffce2afb000)
        libperl.so => /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/x86_64-linux-thread-multi/CORE/libperl.so (0x00007f2366a00000)
        libc.so.6 => /home/linuxbrew/.linuxbrew/lib/libc.so.6 (0x00007f23667ef000)
        libpthread.so.0 => /home/linuxbrew/.linuxbrew/lib/libpthread.so.0 (0x00007f2367034000)
        libdl.so.2 => /home/linuxbrew/.linuxbrew/lib/libdl.so.2 (0x00007f236702f000)
        libm.so.6 => /home/linuxbrew/.linuxbrew/lib/libm.so.6 (0x00007f236670d000)
        libcrypt.so.1 => /home/linuxbrew/.linuxbrew/lib/libcrypt.so.1 (0x00007f2366ff2000)
        /home/linuxbrew/.linuxbrew/lib/ld.so => /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/lib64/ld-linux-x86-64.so.2 (0x00007f236703b000)


xuruizhi@DESKTOP-HI65AUV:~$ bismark
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
        LANGUAGE = (unset),
        LC_ALL = (unset),
        LANG = "C.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
Path to Bowtie 2 specified as: bowtie2
Failed to execute Bowtie porperly (return code of 'bowtie2 --version' was 32512). Please install Bowtie (1/2) first and make sure it is in the PATH, or specify the path to the Bowtie (1/2) using --path_to_bowtie /path/to/bowtie
```

* 解决perl小配置问题  

```bash
问题：
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
    LANGUAGE = (unset),
    LC_ALL = (unset),
    LANG = "C.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").

Path to Bowtie 2 specified as: bowtie2
Failed to execute Bowtie porperly (return code of 'bowtie2 --version' was 32512). Please install Bowtie (1/2) first and make sure it is in the PATH, or specify the path to the Bowtie (1/2) using --path_to_bowtie /path/to/bowtie
```



解决办法：  


[https://blog.csdn.net/zkgoup/article/details/112380842])(https://blog.csdn.net/zkgoup/article/details/112380842)  

```bash
xuruizhi@DESKTOP-HI65AUV:~$ sudo apt-get update
[sudo] password for xuruizhi:
Get:1 https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/ InRelease [3622 B]
。。。
Fetched 3826 kB in 6s (621 kB/s)
Reading package lists... Done


xuruizhi@DESKTOP-HI65AUV:~$ sudo apt-get install  -y language-pack-zh-hans
Reading package lists... Done
Building dependency tree
Reading state information... Done
language-pack-zh-hans is already the newest version (1:20.04+20220818).
0 upgraded, 0 newly installed, 0 to remove and 66 not upgraded.

xuruizhi@DESKTOP-HI65AUV:~$ sudo apt-get install -y language-pack-zh-hant
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  language-pack-zh-hant-base
The following NEW packages will be installed:
  language-pack-zh-hant language-pack-zh-hant-base
0 upgraded, 2 newly installed, 0 to remove and 66 not upgraded.
Need to get 1541 kB of archives.
。。。
Generating locales (this might take a while)...
  zh_HK.UTF-8... done
  zh_TW.UTF-8... done
Generation complete.

xuruizhi@DESKTOP-HI65AUV:~$ locale
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=C.UTF-8
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=

sudo dpkg-reconfigure locales

xuruizhi@DESKTOP-HI65AUV:~$ export LC_ALL=en_US.UTF-8
xuruizhi@DESKTOP-HI65AUV:~$ export LANGUAGE=en_US.UTF-8
xuruizhi@DESKTOP-HI65AUV:~$ export LANG=en_US.UTF-8
xuruizhi@DESKTOP-HI65AUV:/usr/share/locales$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
```

* another问题  

问题:
```bash
xuruizhi@DESKTOP-HI65AUV:~$ bismark
Path to Bowtie 2 specified as: bowtie2
Failed to execute Bowtie porperly (return code of 'bowtie2 --version' was 32512). Please install Bowtie (1/2) first and make sure it is in the PATH, or specify the path to the Bowtie (1/2) using --path_to_bowtie /path/to/bowtie
```

解决办法：
```bash 
# 先安装bowtie2
xuruizhi@DESKTOP-HI65AUV:~$ sudo apt install bowtie2

xuruizhi@DESKTOP-HI65AUV:~$  bowtie2 --version
/usr/bin/bowtie2-align-s version 2.3.5.1
64-bit

xuruizhi@DESKTOP-HI65AUV:~$ bismark
Path to Bowtie 2 specified as: bowtie2
Bowtie seems to be working fine (tested command 'bowtie2 --version' [2.3.5])
Output format is BAM (default)
Alignments will be written out in BAM format. Samtools found here: '/home/linuxbrew/.linuxbrew/bin/samtools'
Genome folder was not specified!


xuruizhi@DESKTOP-HI65AUV:~$ whereis bowtie2
bowtie2: /usr/bin/bowtie2 /usr/share/man/man1/bowtie2.1.gz
xuruizhi@DESKTOP-HI65AUV:~$ export PATH="/usr/bin/bowtie2:$PATH"
```
